parsers/NORDPOOL.py

```diff
@@ -0,0 +1,194 @@
+
+SOURCE = "nordpool.com"
+
+
+class NORDPOOL_API_ENDPOINT(Enum):
+    PRICE = "Auction/Prices/ByAreas"
+
+
+class MARKET_TYPE(Enum):
+    DAY_AHEAD = "DayAhead"
+    GB_DAY_AHEAD = "GbHalfHour_DayAhead"
+
+
+class CURRENCY(Enum):
+    EUR = "EUR"
+    DKK = "DKK"
+    SEK = "SEK"
+    NOK = "NOK"
+    GBP = "GBP"
+
+
+ZONE_MAPPING = {
+    "AT": "AT",
+    "BE": "BE",
+    "DK-DK1": "DK1",
+    "DK-DK2": "DK2",
+    "DE": "GER",
+    "EE": "EE",
+    "FI": "FI",
+    "FR": "FR",
+    "GB": "UK",
+    "LT": "LT",
+    "LV": "LV",
+    "NL": "NL",
+    "NO-NO1": "NO1",
+    "NO-NO2": "NO2",
+    "NO-NO3": "NO3",
+    "NO-NO4": "NO4",
+    "NO-NO5": "NO5",
+    "PL": "PL",
+    "SE-SE1": "SE1",
+    "SE-SE2": "SE2",
+    "SE-SE3": "SE3",
+    "SE-SE4": "SE4",
+}
+
+INVERTED_ZONE_MAPPING = {value: key for key, value in ZONE_MAPPING.items()}
+
+
+def _generate_new_nordpool_token(session: Session) -> NordpoolToken:
+    URL = "https://sts.nordpoolgroup.com/connect/token"
+    username = get_token("EMAPS_NORDPOOL_USERNAME")
+    password = get_token("EMAPS_NORDPOOL_PASSWORD")
+    headers = {
+        "accept": "application/json",
+        "Authorization": "Basic Y2xpZW50X21hcmtldGRhdGFfYXBpOmNsaWVudF9tYXJrZXRkYXRhX2FwaQ==",  # This is a public key listed in the Nordpool API documentation
+        "Content-Type": "application/x-www-form-urlencoded",
+    }
+    data = {
+        "grant_type": "password",
+        "scope": "marketdata_api",
+        "username": username,
+        "password": password,
+    }
+    response = session.post(URL, headers=headers, data=data)
+    response = _handle_status_code(response, getLogger(__name__))
+    token_data = response.json()
+    token = NordpoolToken(
+        token=token_data["access_token"],
+        expiration=datetime.now(tz=timezone.utc)
+        + timedelta(seconds=token_data["expires_in"]),
+    )
+
+    return token
+
+
+def _handle_status_code(response: Response, logger: Logger) -> Response:
+    match response.status_code:
+        case 200:
+            return response
+        case 400:
+            # Bad request
+            raise ValueError(f"Bad request: {response.text}")
+        case 401:
+            # Unauthorized
+            raise PermissionError(f"Unauthorized access: {response.text}")
+        case 403:
+            # Forbidden
+            raise PermissionError(f"Forbidden access: {response.text}")
+        case _:
+            # Other errors
+            raise Exception(
+                f"Unexpected error: Code: {response.status_code}, Text: {response.text}"
+            )
+
+
+# TODO: Remove this when we run on Python 3.11 or above
+def zulu_to_utc(datetime_string: str) -> str:
+    """Converts a zulu time string to a UTC time string."""
+    return datetime_string.replace("Z", "+00:00")
+
+
+def _query_nordpool(
+    endpoint: NORDPOOL_API_ENDPOINT,
+    params: dict[str, str],
+    logger: Logger,
+    session: Session,
+):
+    global CURRENT_TOKEN
```

Comment:
@VIKTORVAV99 a tip here, if you need to update a global variable, you need to reference it with global otherwise you would only be updating the local reference.

Reply:
The previous code was not working because of that.

Reply:
Thanks!

I'm not really used to working with globals in python 😅.